src/middlewared/middlewared/plugins/zfs/property_management.py

```diff
@@ -0,0 +1,167 @@
+    """
+    requested_props = set()
+    is_fs = hdl_type == ZFSType.ZFS_TYPE_FILESYSTEM
+    is_vol = hdl_type == ZFSType.ZFS_TYPE_VOLUME
+    for i in req_props:
+        try:
+            prop = ZFSProperty[i.upper()]
+        except KeyError:
+            # invalid property, nothing to do
+            continue
+
+        if is_fs and prop in property_sets.ZFS_FILESYSTEM_PROPERTIES:
+            requested_props.add(prop)
+        elif is_vol and prop in property_sets.ZFS_VOLUME_PROPERTIES:
+            requested_props.add(prop)
+
+        if prop in PROPERTY_TEMPLATES.crypto:
+            # if any property requested is a crypto property,
+            # we're going to include the other crypto related
+            # properties for the convenience of the api user
+            requested_props |= PROPERTY_TEMPLATES.crypto
+
+    # Cache the result for this type
+    requested_props = requested_props or set()
+    if hdl_type == ZFSType.ZFS_TYPE_FILESYSTEM:
+        det_props.fs = requested_props
+    elif hdl_type == ZFSType.ZFS_TYPE_VOLUME:
+        det_props.vol = requested_props
+
+    return frozenset(requested_props)
+
+
+def build_set_of_zfs_props(
+    hdl_type: ZFSResourceType,
+    det_props: DeterminedProperties,
+    req_props: list[str] | None,
+) -> frozenset["ZFSProperty"]:
+    """Build a set of ZFS properties to retrieve for a given ZFS resource.
+
+    This function determines which ZFS properties should be retrieved based on
+    the resource type and requested properties. It uses caching to avoid
+    recomputing property sets for the same resource types.
+
+    Args:
+        hdl_type: The ZFS resource type (filesystem, volume, or snapshot)
+        det_props: DeterminedProperties instance used for caching results
+        req_props: List of requested property names as strings, or None
+            for no properties. An empty list (default) requests space
+            related properties.
+
+    Returns:
+        frozenset[ZFSProperty] | None: Set of valid ZFS properties to retrieve,
+            default properties if req_props is None, or None if no properties
+            should be retrieved (empty req_props list or unsupported type).
+    """
+    if req_props is None:
+        # If the req_props (requested properties) is None, then
+        # the user explicitly requested no zfs properties be retrieved
+        return frozenset()
+    elif req_props == [] or hdl_type not in (
+        ZFSType.ZFS_TYPE_FILESYSTEM,
+        ZFSType.ZFS_TYPE_VOLUME,
+    ):
+        # No properties were requested, or unsupported resource type
+        return PROPERTY_TEMPLATES.default
+
+    # Check if we already have cached properties for this type
+    if hdl_type == ZFSType.ZFS_TYPE_FILESYSTEM and det_props.fs is not None:
+        return det_props.fs
+    elif hdl_type == ZFSType.ZFS_TYPE_VOLUME and det_props.vol is not None:
+        return det_props.vol
+
+    # Now cache and return the zfs properties for this type
```

Comment:
Sorry, I don't seem to understand, where does caching occur here?

Reply:
`det_props` is an instance of `DeterminedProperties`. __build_cache caches the zfs properties on the class